core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -1413,6 +1436,41 @@ object ZManaged {
    */
   val never: ZManaged[Any, Nothing, Nothing] = ZManaged.fromEffect(ZIO.never)
 
+  /**
+   * Creates a scope in which resources can be safely preallocated.
+   */
+  final def scope[R]: ZManaged[R, Nothing, Preallocate] =
+    ZManaged {
+      Ref.make(Map.empty[Int, Exit[Any, Any] => ZIO[R, Nothing, Any]]).flatMap { finalizers =>
+        Ref.make(-1).map(_.update(_ + 1)).map { nextIndex =>
+          Reservation(
+            acquire = ZIO.succeed {
+              new Preallocate {
+                override def apply[R, E, A](managed: ZManaged[R, E, A]) =
+                  ZIO.uninterruptibleMask { restore =>
+                    for {
+                      env      <- ZIO.environment[R]
+                      res      <- managed.reserve
+                      resource <- restore(res.acquire).onError(err => res.release(Exit.Failure(err)))
+                      index    <- nextIndex
+                      release  = res.release.andThen(_.provide(env))
+                      _        <- finalizers.update(_ + (index -> release))
+                    } yield ZManaged
+                      .make(ZIO.succeed(resource))(_ => release(Exit.Success(resource)) *> finalizers.update(_ - index))
```

Comment:
```suggestion
                      .make(ZIO.succeed(resource))(_ => release(Exit.Success(resource)).ensuring(finalizers.update(_ - index)))
```

to avoid double execution if a defect occurs?

Reply:
good catch!